File: web/frontend/src/components/project/soap/SoapOperation.js
/*
 Copyright 2020 Karl Dahlgren

 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 */

import React, {PureComponent} from 'react';
import {Link} from "react-router-dom";
import axios from "axios";
import ToolkitProvider, {Search} from "react-bootstrap-table2-toolkit";
import BootstrapTable from "react-bootstrap-table-next";
import PaginationFactory from "react-bootstrap-table2-paginator";
import {connect} from "react-redux";
import {setAuthenticationState} from "../../../redux/Actions";
import validateErrorResponse from "../../../utility/HttpResponseValidator";
const { SearchBar } = Search;

const SELECT = true;
const DESELECT = false;

class SoapOperation extends PureComponent {

    constructor(props) {
        super(props);

        this.onRowSelect = this.onRowSelect.bind(this);
        this.onRowSelectAll = this.onRowSelectAll.bind(this);
        this.nameFormat = this.nameFormat.bind(this);
        this.setNewMockResponseName = this.setNewMockResponseName.bind(this);
        this.setNewMockResponseStatus = this.setNewMockResponseStatus.bind(this);
        this.onCreateMockResponseClick = this.onCreateMockResponseClick.bind(this);
        this.onDeleteMockResponsesClick = this.onDeleteMockResponsesClick.bind(this);

        this.columns = [{
            dataField: 'id',
            text: 'id',
            hidden: true
        }, {
            dataField: 'name',
            text: 'Response name',
            sort: true,
            formatter: this.nameFormat
        }, {
            dataField: 'status',
            text: 'Status',
            sort: true
        }, {
            dataField: 'httpStatusCode',
            text: 'HTTP status code',
            sort: true
        }];

        this.updateColumns = [{
            dataField: 'id',
            text: 'id',
            hidden: true
        }, {
            dataField: 'name',
            text: 'Name',
            sort: true
        }];

        this.selectRow = {
            mode: 'checkbox',
            onSelect: this.onRowSelect,
            onSelectAll: this.onRowSelectAll
        };

        this.defaultSort = [{
            dataField: 'name',
            order: 'asc'
        }];

        this.state = {
            projectId: this.props.match.params.projectId,
            portId: this.props.match.params.portId,
            operationId: this.props.match.params.operationId,
            operation: {
                operationIdentifier: {},
                mockResponses: []
            },
            selectedMockResponses: [],
            newMockResponse: {
                name: "",
                status: "ENABLED"
            }
        };

        this.getOperation();
    }

    setNewMockResponseName(name) {
        this.setState({ newMockResponse: {
                ...this.state.newMockResponse,
                name: name
            }
        });
    }

    setNewMockResponseStatus(status) {
        this.setState({ newMockResponse: {
                ...this.state.newMockResponse,
                status: status
            }
        });
    }

    onRowSelect(value, mode) {
        let mockResponses = this.state.selectedMockResponses.slice();
        let mockResponse = {
            id: value.id,
            name: value.name
        };
        if(mode === SELECT){
            mockResponses.push(mockResponse);
        } else if(mode === DESELECT){
            let index = mockResponses.indexOf(mockResponse);
            mockResponses.splice(index, 1);
        }
        this.setState({
            selectedMockResponses: mockResponses
        });
    }

    onRowSelectAll(mode) {
        if(mode === SELECT){
            let mockResponses = [];
            this.state.operation.mockResponses.forEach(value => {
                let mockResponse = {
                    id: value.id,
                    name: value.name
                };
                mockResponses.push(mockResponse);
            });
            this.setState({
                selectedMockResponses: mockResponses
            });
        } else if(mode === DESELECT){
            this.setState({
                selectedMockResponses: []
            });
        }
    }

    nameFormat(cell, row) {
        if(cell == null){
            return;
        }

        return (
            <div className="table-link">
                <Link to={"/web/soap/project/" + this.state.projectId + "/port/" + this.state.portId + "/operation/" + this.state.operationId + "/mockresponse/" + row.id }>{cell}</Link>
            </div>
        )
    }

    getOperation() {
        axios
            .get("/api/rest/soap/project/" + this.state.projectId + "/port/" + this.state.portId + "/operation/" + this.state.operationId)
            .then(response => {
                this.setState({
                    operation: response.data,
                });
            })
            .catch(error => {
                validateErrorResponse(error, this.props.setAuthenticationState)
            });
    }

    onCreateMockResponseClick() {
        axios
            .post("/api/rest/soap/project/" + this.state.projectId + "/port/" +
                this.state.portId + "/operation/" + this.state.operationId + "/response",
                this.state.newMockResponse)
            .then(response => {
                this.props.history.push("/web/soap/project/" + this.state.projectId + "/port/" + this.state.portId + "/operation/" + this.state.operationId + "/response");
            })
            .catch(error => {
                validateErrorResponse(error, this.props.setAuthenticationState)
            });
    }

    onDeleteMockResponsesClick() {
        this.state.selectedMockResponses.forEach(mockResponse => {
            axios
                .delete("/api/rest/soap/project/" + this.state.projectId + "/port/" +
                    this.state.portId + "/operation/" + this.state.operationId + "/response/" + mockResponse.id)
                .then(response => {
                    this.getOperation();
                })
                .catch(error => {
                    validateErrorResponse(error, this.props.setAuthenticationState)
                });
        })

    }

    render() {
        return (
            <div>
                <section>
                    <div className="navigation">
                        <nav aria-label="breadcrumb">
                            <ol className="breadcrumb breadcrumb-custom">
                                <li className="breadcrumb-item"><Link to={"/web"}>Home</Link></li>
                                <li className="breadcrumb-item"><Link to={"/web/soap/project/" + this.state.projectId}>Project</Link></li>
                                <li className="breadcrumb-item"><Link to={"/web/soap/project/" + this.state.projectId + "/port/" + this.state.portId}>Port</Link></li>
                                <li className="breadcrumb-item">{this.state.operation.name}</li>
                            </ol>
                        </nav>
                    </div>
                    <div className="content-top">
                        <div className="title">
                            <h1>Operation: {this.state.operation.name}</h1>
                        </div>
                        <div className="menu" align="right">
                            <button className="btn btn-success demo-button-disabled menu-button" data-toggle="modal" data-target="#updateOperationModal"><i className="fas fa-plus-circle"/> <span>Update operation</span></button>
                            <button className="btn btn-primary demo-button-disabled menu-button" data-toggle="modal" data-target="#createMockResponseModal"><i className="fas fa-plus-circle"/> <span>Create response</span></button>
                        </div>
                    </div>
                    <div className="content-summary">
                        <dl className="row">
                            <dt className="col-sm-3 content-title">Identifier</dt>
                            <dd className="col-sm-9">{this.state.operation.operationIdentifier.name}</dd>
                        </dl>
                        <dl className="row">
                            <dt className="col-sm-3 content-title">SOAP Version</dt>
                            <dd className="col-sm-9">{this.state.operation.soapVersion}</dd>
                        </dl>
                        <dl className="row">
                            <dt className="col-sm-3 content-title">Identify strategy</dt>
                            <dd className="col-sm-9">{this.state.operation.identifyStrategy}</dd>
                        </dl>
                        <dl className="row">
                            <dt className="col-sm-3 content-title">Status</dt>
                            <dd className="col-sm-9">{this.state.operation.status}</dd>
                        </dl>
                        <dl className="row">
                            <dt className="col-sm-3 content-title">Response strategy</dt>
                            <dd className="col-sm-9">{this.state.operation.responseStrategy}</dd>
                        </dl>
                        <dl className="row">
                            <dt className="col-sm-3 content-title">Address</dt>
                            <dd className="col-sm-9">Test</dd>
                        </dl>
                        <dl className="row">
                            <dt className="col-sm-3 content-title">Original endpoint</dt>
                            <dd className="col-sm-9">{this.state.operation.originalEndpoint}</dd>
                        </dl>
                        <dl className="row">
                            <dt className="col-sm-3 content-title">Forwarded endpoint</dt>
                            <dd className="col-sm-9">{this.state.operation.forwardedEndpoint}</dd>
                        </dl>
                        <dl className="row">
                            <dt className="col-sm-3 content-title">Simulate network delay</dt>
                            <dd className="col-sm-9">{this.state.operation.simulateNetworkDelay}</dd>
                        </dl>
                        <dl className="row">
                            <dt className="col-sm-3 content-title">Network delay</dt>
                            <dd className="col-sm-9">{this.state.operation.networkDelay}</dd>
                        </dl>
                        <dl className="row">
                            <dt className="col-sm-3 content-title">Default response</dt>
                            <dd className="col-sm-9">{this.state.operation.defaultResponseName}</dd>
                        </dl>
                        <dl className="row">
                            <dt className="col-sm-3 content-title">Mock on failure</dt>
                            <dd className="col-sm-9">{this.state.operation.mockOnFailure}</dd>
                        </dl>
                    </div>
                    <div className="panel panel-primary table-panel">
                        <div className="panel-heading table-panel-heading">
                            <h3 className="panel-title">Mock responses</h3>
                        </div>
                        <div className="table-result">
                            <ToolkitProvider bootstrap4
                                             columns={ this.columns}
                                             data={this.state.operation.mockResponses}
                                             keyField="id"
                                             search>
                                {
                                    (props) => (
                                        <div>
                                            <div>
                                                <SearchBar {...props.searchProps} className={"table-filter-field"}/>
                                            </div>
                                            <BootstrapTable {...props.baseProps} bootstrap4
                                                            data={this.state.operation.mockResponses} columns={this.columns}
                                                            defaultSorted={this.defaultSort} keyField='id' hover
                                                            selectRow={this.selectRow}
                                                            pagination={ PaginationFactory() }/>
                                        </div>
                                    )}
                            </ToolkitProvider>
                            <div className="panel-buttons">
                                <button className="btn btn-primary demo-button-disabled menu-button" data-toggle="modal"
                                        disabled={this.state.selectedMockResponses.length === 0}
                                        data-target="#updateStatusModal"><i className="fas fa-plus-circle"/> <span>Update status</span></button>
                                <button className="btn btn-primary demo-button-disabled menu-button" data-toggle="modal"
                                        disabled={this.state.selectedMockResponses.length === 0}
                                        data-target="#updateOperationModal"><i className="fas fa-plus-circle"/> <span>Duplicate</span></button>
                                <button className="btn btn-danger demo-button-disabled menu-button" data-toggle="modal"
                                        disabled={this.state.selectedMockResponses.length === 0}
                                        data-target="#deleteMockResponsesModal"><i className="fas fa-plus-circle"/> <span>Delete responses</span></button>
                            </div>
                        </div>
                    </div>
                </section>

                <div className="modal fade" id="updateOperationModal" tabIndex="-1" role="dialog"
                     aria-labelledby="updateOperationModalLabel" aria-hidden="true">
                    <div className="modal-dialog modal-dialog-centered modal-lg" role="document">
                        <div className="modal-content">
                            <div className="modal-header">
                                <h5 className="modal-title" id="updateOperationModalLabel">Update operation</h5>
                                <button type="button" className="close" data-dismiss="modal" aria-label="Close">
                                    <span aria-hidden="true">&times;</span>
                                </button>
                            </div>
                            <div className="modal-body">
                                <form>
                                    <div className="form-group row">
                                        <label htmlFor="newOperationStatus" className="col-sm-3 col-form-label">Status</label>
                                        <div className="col-sm-9">
                                            <select id="inputStatus" className="form-control" defaultValue={this.state.operation.status}>
                                                <option>MOCKED</option>
                                                <option>DISABLED</option>
                                                <option>FORWARDED</option>
                                                <option>RECORDING</option>
                                                <option>RECORD_ONCE</option>
                                                <option>ECHO</option>
                                            </select>
                                        </div>
                                    </div>
                                    <div className="form-group row">
                                        <label htmlFor="newOperationResponseStrategy" className="col-sm-3 col-form-label">Response strategy</label>
                                        <div className="col-sm-9">
                                            <select id="inputStatus" className="form-control" defaultValue={this.state.operation.responseStrategy}>
                                                <option>RANDOM</option>
                                                <option>SEQUENCE</option>
                                                <option>XPATH_INPUT</option>
                                            </select>
                                        </div>
                                    </div>
                                    <div className="form-group row">
                                        <label className="col-sm-3 col-form-label">Response strategy</label>
                                        <div className="col-sm-9">
                                            <input className="form-control" type="text" />
                                        </div>
                                    </div>
                                    <div className="form-group row">
                                        <label className="col-sm-3 col-form-label">Simulate network delay</label>
                                        <div className="col-sm-9">
                                            <input type="checkbox" />
                                        </div>
                                    </div>
                                    <div className="form-group row">
                                        <label className="col-sm-3 col-form-label">Network delay</label>
                                        <div className="col-sm-9">
                                            <input className="form-control" type="text" />
                                        </div>
                                    </div>
                                    <div className="form-group row">
                                        <label htmlFor="newOperationResponseStrategy" className="col-sm-3 col-form-label">Default response</label>
                                        <div className="col-sm-9">
                                            <select id="inputStatus" className="form-control" >

                                            </select>
                                        </div>
                                    </div>
                                    <div className="form-group row">
                                        <label className="col-sm-3 col-form-label">Mock on failure</label>
                                        <div className="col-sm-9">
                                            <input type="checkbox" />
                                        </div>
                                    </div>
                                </form>
                            </div>
                            <div className="modal-footer">
                                <button className="btn btn-success" data-dismiss="modal" onClick={this.onExportOperationsClick}>Update</button>
                            </div>
                        </div>
                    </div>
                </div>

                <div className="modal fade" id="createMockResponseModal" tabIndex="-1" role="dialog"
                     aria-labelledby="createMockResponseModalLabel" aria-hidden="true">
                    <div className="modal-dialog modal-dialog-centered modal-lg" role="document">
                        <div className="modal-content">
                            <div className="modal-header">
                                <h5 className="modal-title" id="createMockResponseModalLabel">Create response?</h5>
                                <button type="button" className="close" data-dismiss="modal" aria-label="Close">
                                    <span aria-hidden="true">&times;</span>
                                </button>
                            </div>
                            <div className="modal-body">
                                <div className="form-group row">
                                    <label className="col-sm-2 col-form-label">Name</label>
                                    <div className="col-sm-10">
                                        <input className="form-control" type="text" onChange={event => this.setNewMockResponseName(event.target.value)} />
                                    </div>
                                </div>
                                <div className="form-group row">
                                    <label className="col-sm-2 col-form-label">Status</label>
                                    <div className="col-sm-10">
                                        <select id="inputStatus" className="form-control" defaultValue="MOCKED" onChange={event => this.setNewMockResponseStatus(event.target.value)}>
                                            <option>ENABLED</option>
                                            <option>DISABLED</option>
                                        </select>
                                    </div>
                                </div>
                            </div>
                            <div className="modal-footer">
                                <button className="btn btn-primary" data-dismiss="modal" onClick={this.onCreateMockResponseClick}>Create</button>
                            </div>
                        </div>
                    </div>
                </div>
                
                <div className="modal fade" id="updateStatusModal" tabIndex="-1" role="dialog"
                     aria-labelledby="updateStatusModalLabel" aria-hidden="true">
                    <div className="modal-dialog modal-dialog-centered modal-lg" role="document">
                        <div className="modal-content">
                            <div className="modal-header">
                                <h5 className="modal-title" id="updateStatusModalLabel">Update status?</h5>
                                <button type="button" className="close" data-dismiss="modal" aria-label="Close">
                                    <span aria-hidden="true">&times;</span>
                                </button>
                            </div>
                            <div className="modal-body">
                                <p>Do you want update the status for the following mock responses?</p>
                                <div className="table-result">
                                    <ToolkitProvider bootstrap4
                                                     columns={ this.updateColumns}
                                                     data={ this.state.selectedMockResponses }
                                                     keyField="id">
                                        {
                                            (props) => (
                                                <div>
                                                    <BootstrapTable { ...props.baseProps } bootstrap4 data={this.state.selectedMockResponses} columns={this.updateColumns}
                                                                    defaultSorted={ this.defaultSort } keyField='id' hover
                                                                    striped
                                                                    pagination={ PaginationFactory({hideSizePerPage: true}) }/>
                                                </div>
                                            )}
                                    </ToolkitProvider>
                                </div>
                                <div className="form-group row">
                                    <label className="col-sm-2 col-form-label">Status</label>
                                    <div className="col-sm-10">
                                        <select id="inputStatus" className="form-control" defaultValue="MOCKED">
                                            <option>MOCKED</option>
                                            <option>DISABLED</option>
                                            <option>FORWARDED</option>
                                            <option>RECORDING</option>
                                            <option>RECORD_ONCE</option>
                                            <option>ECHO</option>
                                        </select>
                                    </div>
                                </div>
                            </div>
                            <div className="modal-footer">
                                <button className="btn btn-primary" data-dismiss="modal">Update</button>
                            </div>
                        </div>
                    </div>
                </div>

                <div className="modal fade" id="updateEndpointModal" tabIndex="-1" role="dialog"
                     aria-labelledby="updateEndpointModalLabel" aria-hidden="true">
                    <div className="modal-dialog modal-dialog-centered modal-lg" role="document">
                        <div className="modal-content">
                            <div className="modal-header">
                                <h5 className="modal-title" id="updateEndpointModalLabel">Update endpoint?</h5>
                                <button type="button" className="close" data-dismiss="modal" aria-label="Close">
                                    <span aria-hidden="true">&times;</span>
                                </button>
                            </div>
                            <div className="modal-body">
                                <p>Do you want update the endpoint for the following mockResponses?</p>
                                <div className="table-result">
                                    <ToolkitProvider bootstrap4
                                                     columns={ this.updateColumns}
                                                     data={ this.state.selectedMockResponses }
                                                     keyField="id">
                                        {
                                            (props) => (
                                                <div>
                                                    <BootstrapTable { ...props.baseProps } bootstrap4 data={this.state.selectedMockResponses} columns={this.updateColumns}
                                                                    defaultSorted={ this.defaultSort } keyField='id' hover
                                                                    striped
                                                                    pagination={ PaginationFactory({hideSizePerPage: true}) }/>
                                                </div>
                                            )}
                                    </ToolkitProvider>
                                </div>
                                <div className="form-group row">
                                    <label className="col-sm-2 col-form-label">Endpoint</label>
                                    <div className="col-sm-10">
                                        <input className="form-control" type="text" />
                                    </div>
                                </div>
                            </div>
                            <div className="modal-footer">
                                <button className="btn btn-primary" data-dismiss="modal">Update</button>
                            </div>
                        </div>
                    </div>
                </div>

                <div className="modal fade" id="deleteMockResponsesModal" tabIndex="-1" role="dialog"
                     aria-labelledby="deleteMockResponsesModalLabel" aria-hidden="true">
                    <div className="modal-dialog modal-dialog-centered modal-lg" role="document">
                        <div className="modal-content">
                            <div className="modal-header">
                                <h5 className="modal-title" id="deleteMockResponsesModalLabel">Delete mockResponses?</h5>
                                <button type="button" className="close" data-dismiss="modal" aria-label="Close">
                                    <span aria-hidden="true">&times;</span>
                                </button>
                            </div>
                            <div className="modal-body">
                                <p>Do you want delete the following mockResponses?</p>
                                <div className="table-result">
                                    <ToolkitProvider bootstrap4
                                                     columns={ this.updateColumns}
                                                     data={ this.state.selectedMockResponses }
                                                     keyField="id">
                                        {
                                            (props) => (
                                                <div>
                                                    <BootstrapTable { ...props.baseProps } bootstrap4 data={this.state.selectedMockResponses} columns={this.updateColumns}
                                                                    defaultSorted={ this.defaultSort } keyField='id' hover
                                                                    striped
                                                                    pagination={ PaginationFactory({hideSizePerPage: true}) }/>
                                                </div>
                                            )}
                                    </ToolkitProvider>
                                </div>
                            </div>
                            <div className="modal-footer">
                                <button className="btn btn-danger" data-dismiss="modal" onClick={this.onDeleteMockResponsesClick}>Delete</button>
                            </div>
                        </div>
                    </div>
                </div>
            </div>
        )
    }
}

export default connect(
    null,
    { setAuthenticationState }
)(SoapOperation);

File: model/model-mock/model-mock-soap/src/main/java/com/castlemock/model/mock/soap/domain/SoapOperation.java
/*
 * Copyright 2015 Karl Dahlgren
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.castlemock.model.mock.soap.domain;

import com.castlemock.model.core.http.HttpMethod;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import com.fasterxml.jackson.databind.annotation.JsonPOJOBuilder;

import javax.xml.bind.annotation.XmlAccessType;
import javax.xml.bind.annotation.XmlAccessorType;
import javax.xml.bind.annotation.XmlElement;
import javax.xml.bind.annotation.XmlElementWrapper;
import javax.xml.bind.annotation.XmlRootElement;
import java.util.List;
import java.util.Objects;
import java.util.Optional;
import java.util.stream.Collectors;

/**
 * @author Karl Dahlgren
 * @since 1.0
 */
@XmlRootElement
@XmlAccessorType(XmlAccessType.NONE)
@JsonDeserialize(builder = SoapOperation.Builder.class)
public class SoapOperation {

    @XmlElement
    private final String id;

    @XmlElement
    private final String name;

    @XmlElement
    private final String identifier;

    @XmlElement
    private final SoapOperationIdentifier operationIdentifier;

    @XmlElement
    private final SoapResponseStrategy responseStrategy;

    @XmlElement
    private final SoapOperationStatus status;

    @XmlElement
    private final HttpMethod httpMethod;

    @XmlElement
    private final SoapVersion soapVersion;

    @XmlElement
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    private final String defaultBody;

    @XmlElement
    private final Integer currentResponseSequenceIndex;

    @XmlElement
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    private final String forwardedEndpoint;

    @XmlElement
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    private final String originalEndpoint;

    @XmlElement
    private final Boolean simulateNetworkDelay;

    @XmlElement
    private final Long networkDelay;

    @XmlElement
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    private final String defaultMockResponseId;

    @XmlElement
    private final String portId;

    @XmlElement
    private final Boolean mockOnFailure;

    @XmlElement
    private final SoapOperationIdentifyStrategy identifyStrategy;

    @XmlElementWrapper(name = "mockResponses")
    @XmlElement(name = "mockResponse")
    private final List<SoapMockResponse> mockResponses;

    @XmlElement
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    private final String invokeAddress;

    @XmlElement
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    private final String defaultResponseName;

    @XmlElement
    private final Boolean automaticForward;


    private SoapOperation(final Builder builder){
        this.id = Objects.requireNonNull(builder.id, "id");
        this.name = Objects.requireNonNull(builder.name, "name");
        this.identifier = Objects.requireNonNull(builder.identifier, "identifier");
        this.operationIdentifier = Objects.requireNonNull(builder.operationIdentifier, "operationIdentifier");
        this.responseStrategy = Objects.requireNonNull(builder.responseStrategy, "responseStrategy");
        this.status = Objects.requireNonNull(builder.status, "status");
        this.httpMethod = Objects.requireNonNull(builder.httpMethod, "httpMethod");
        this.soapVersion = Objects.requireNonNull(builder.soapVersion, "soapVersion");
        this.currentResponseSequenceIndex = Objects.requireNonNull(builder.currentResponseSequenceIndex, "currentResponseSequenceIndex");
        this.portId = Objects.requireNonNull(builder.portId, "portId");
        this.identifyStrategy = Objects.requireNonNull(builder.identifyStrategy, "identifyStrategy");

        this.simulateNetworkDelay = builder.simulateNetworkDelay;
        this.defaultBody = builder.defaultBody;
        this.originalEndpoint = builder.originalEndpoint;
        this.forwardedEndpoint = builder.forwardedEndpoint;
        this.networkDelay = builder.networkDelay;
        this.defaultMockResponseId = builder.defaultMockResponseId;
        this.invokeAddress = builder.invokeAddress;
        this.defaultResponseName = builder.defaultResponseName;
        this.mockOnFailure = builder.mockOnFailure;
        this.automaticForward = builder.automaticForward;
        this.mockResponses = Optional.ofNullable(builder.mockResponses).orElseGet(List::of);
    }

    public String getId() {
        return id;
    }

    public String getName() {
        return name;
    }

    public String getIdentifier() {
        return identifier;
    }

    public SoapOperationIdentifier getOperationIdentifier() {
        return operationIdentifier;
    }

    public String getPortId() {
        return portId;
    }

    public SoapResponseStrategy getResponseStrategy() {
        return responseStrategy;
    }

    public SoapOperationStatus getStatus() {
        return status;
    }


    public List<SoapMockResponse> getMockResponses() {
        return Optional.ofNullable(mockResponses)
                .map(List::copyOf)
                .orElseGet(List::of);
    }

    public Optional<String> getInvokeAddress() {
        return Optional.ofNullable(invokeAddress);
    }

    public Optional<String> getDefaultBody() {
        return Optional.ofNullable(defaultBody);
    }

    public HttpMethod getHttpMethod() {
        return httpMethod;
    }

    public SoapVersion getSoapVersion() {
        return soapVersion;
    }

    public Integer getCurrentResponseSequenceIndex() {
        return currentResponseSequenceIndex;
    }

    public Optional<String> getForwardedEndpoint() {
        return Optional.ofNullable(forwardedEndpoint);
    }

    public Optional<String> getOriginalEndpoint() {
        return Optional.ofNullable(originalEndpoint);
    }

    public Optional<Boolean> getSimulateNetworkDelay() {
        return Optional.ofNullable(simulateNetworkDelay);
    }

    public Optional<Long> getNetworkDelay() {
        return Optional.ofNullable(networkDelay);
    }

    public Optional<String> getDefaultResponseName() {
        return Optional.ofNullable(defaultResponseName);
    }

    public Optional<Boolean> getMockOnFailure() {
        return Optional.ofNullable(mockOnFailure);
    }

    public SoapOperationIdentifyStrategy getIdentifyStrategy() {
        return identifyStrategy;
    }

    public Optional<String> getDefaultMockResponseId() {
        return Optional.ofNullable(defaultMockResponseId);
    }

    public Optional<Boolean> getAutomaticForward() {
        return Optional.ofNullable(automaticForward);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        SoapOperation that = (SoapOperation) o;
        return automaticForward == that.automaticForward && Objects.equals(id, that.id)
                && Objects.equals(name, that.name) && Objects.equals(identifier, that.identifier)
                && Objects.equals(operationIdentifier, that.operationIdentifier) && responseStrategy == that.responseStrategy
                && status == that.status && httpMethod == that.httpMethod && soapVersion == that.soapVersion
                && Objects.equals(defaultBody, that.defaultBody) && Objects.equals(currentResponseSequenceIndex, that.currentResponseSequenceIndex)
                && Objects.equals(forwardedEndpoint, that.forwardedEndpoint) && Objects.equals(originalEndpoint, that.originalEndpoint)
                && Objects.equals(simulateNetworkDelay, that.simulateNetworkDelay) && Objects.equals(networkDelay, that.networkDelay)
                && Objects.equals(defaultMockResponseId, that.defaultMockResponseId) && Objects.equals(portId, that.portId)
                && Objects.equals(mockOnFailure, that.mockOnFailure) && identifyStrategy == that.identifyStrategy
                && Objects.equals(mockResponses, that.mockResponses) && Objects.equals(invokeAddress, that.invokeAddress)
                && Objects.equals(defaultResponseName, that.defaultResponseName);
    }

    @Override
    public int hashCode() {
        return Objects.hash(id, name, identifier, operationIdentifier, responseStrategy, status, httpMethod,
                soapVersion, defaultBody, currentResponseSequenceIndex, forwardedEndpoint, originalEndpoint,
                simulateNetworkDelay, networkDelay, defaultMockResponseId, portId, mockOnFailure, identifyStrategy,
                mockResponses, invokeAddress, defaultResponseName, automaticForward);
    }

    @Override
    public String toString() {
        return "SoapOperation{" +
                "id='" + id + '\'' +
                ", name='" + name + '\'' +
                ", identifier='" + identifier + '\'' +
                ", operationIdentifier=" + operationIdentifier +
                ", responseStrategy=" + responseStrategy +
                ", status=" + status +
                ", httpMethod=" + httpMethod +
                ", soapVersion=" + soapVersion +
                ", defaultBody='" + defaultBody + '\'' +
                ", currentResponseSequenceIndex=" + currentResponseSequenceIndex +
                ", forwardedEndpoint='" + forwardedEndpoint + '\'' +
                ", originalEndpoint='" + originalEndpoint + '\'' +
                ", simulateNetworkDelay=" + simulateNetworkDelay +
                ", networkDelay=" + networkDelay +
                ", defaultMockResponseId='" + defaultMockResponseId + '\'' +
                ", portId='" + portId + '\'' +
                ", mockOnFailure=" + mockOnFailure +
                ", identifyStrategy=" + identifyStrategy +
                ", mockResponses=" + mockResponses +
                ", invokeAddress='" + invokeAddress + '\'' +
                ", defaultResponseName='" + defaultResponseName + '\'' +
                ", automaticForward=" + automaticForward +
                '}';
    }

    public static Builder builder() {
        return new Builder();
    }

    public Builder toBuilder() {
        return builder()
                .id(id)
                .automaticForward(automaticForward)
                .mockOnFailure(mockOnFailure)
                .portId(portId)
                .simulateNetworkDelay(simulateNetworkDelay)
                .identifyStrategy(identifyStrategy)
                .currentResponseSequenceIndex(currentResponseSequenceIndex)
                .defaultBody(defaultBody)
                .defaultResponseName(defaultResponseName)
                .defaultMockResponseId(defaultMockResponseId)
                .name(name)
                .identifier(identifier)
                .operationIdentifier(operationIdentifier)
                .responseStrategy(responseStrategy)
                .invokeAddress(invokeAddress)
                .soapVersion(soapVersion)
                .forwardedEndpoint(forwardedEndpoint)
                .mockResponses(Optional.ofNullable(mockResponses)
                        .map(responses -> responses.stream()
                                .map(SoapMockResponse::toBuilder)
                                .map(SoapMockResponse.Builder::build)
                                .collect(Collectors.toList()))
                        .orElse(null))
                .originalEndpoint(originalEndpoint)
                .status(status)
                .httpMethod(httpMethod)
                .networkDelay(networkDelay);
    }

    @JsonPOJOBuilder(withPrefix = "")
    public static final class Builder {
        private String id;
        private String name;
        private String identifier;
        private SoapOperationIdentifier operationIdentifier;
        private SoapResponseStrategy responseStrategy;
        private SoapOperationStatus status;
        private HttpMethod httpMethod;
        private SoapVersion soapVersion;
        private String defaultBody;
        private Integer currentResponseSequenceIndex;
        private String forwardedEndpoint;
        private String originalEndpoint;
        private Boolean simulateNetworkDelay;
        private Long networkDelay;
        private String defaultMockResponseId;
        private String portId;
        private Boolean mockOnFailure;
        private SoapOperationIdentifyStrategy identifyStrategy;
        private List<SoapMockResponse> mockResponses;
        private String invokeAddress;
        private String defaultResponseName;
        private Boolean automaticForward;

        private Builder() {
        }

        public Builder id(final String id) {
            this.id = id;
            return this;
        }

        public Builder name(final String name) {
            this.name = name;
            return this;
        }

        public Builder identifier(final String identifier) {
            this.identifier = identifier;
            return this;
        }

        public Builder operationIdentifier(final SoapOperationIdentifier operationIdentifier) {
            this.operationIdentifier = operationIdentifier;
            return this;
        }

        public Builder responseStrategy(final SoapResponseStrategy responseStrategy) {
            this.responseStrategy = responseStrategy;
            return this;
        }

        public Builder status(final SoapOperationStatus status) {
            this.status = status;
            return this;
        }

        public Builder httpMethod(final HttpMethod httpMethod) {
            this.httpMethod = httpMethod;
            return this;
        }

        public Builder soapVersion(final SoapVersion soapVersion) {
            this.soapVersion = soapVersion;
            return this;
        }

        public Builder defaultBody(final String defaultBody) {
            this.defaultBody = defaultBody;
            return this;
        }

        public Builder currentResponseSequenceIndex(final Integer currentResponseSequenceIndex) {
            this.currentResponseSequenceIndex = currentResponseSequenceIndex;
            return this;
        }

        public Builder forwardedEndpoint(final String forwardedEndpoint) {
            this.forwardedEndpoint = forwardedEndpoint;
            return this;
        }

        public Builder originalEndpoint(final String originalEndpoint) {
            this.originalEndpoint = originalEndpoint;
            return this;
        }

        public Builder simulateNetworkDelay(final Boolean simulateNetworkDelay) {
            this.simulateNetworkDelay = simulateNetworkDelay;
            return this;
        }

        public Builder networkDelay(final Long networkDelay) {
            this.networkDelay = networkDelay;
            return this;
        }

        public Builder defaultMockResponseId(final String defaultMockResponseId) {
            this.defaultMockResponseId = defaultMockResponseId;
            return this;
        }

        public Builder portId(final String portId) {
            this.portId = portId;
            return this;
        }

        public Builder mockOnFailure(final Boolean mockOnFailure) {
            this.mockOnFailure = mockOnFailure;
            return this;
        }

        public Builder identifyStrategy(final SoapOperationIdentifyStrategy identifyStrategy) {
            this.identifyStrategy = identifyStrategy;
            return this;
        }

        public Builder mockResponses(final List<SoapMockResponse> mockResponses) {
            this.mockResponses = mockResponses;
            return this;
        }

        public Builder invokeAddress(final String invokeAddress) {
            this.invokeAddress = invokeAddress;
            return this;
        }

        public Builder defaultResponseName(final String defaultResponseName) {
            this.defaultResponseName = defaultResponseName;
            return this;
        }

        public Builder automaticForward(final Boolean automaticForward) {
            this.automaticForward = automaticForward;
            return this;
        }

        public SoapOperation build() {
            return new SoapOperation(this);
        }
    }
}
